src/executor.rs

```diff
@@ -199,3 +133,96 @@ impl Executor {
         Ok(())
     }
 }
+
+async fn task_result(
+    task: &mut Option<JoinHandle<Result<Option<ExitStatus>>>>,
+) -> Result<Option<ExitStatus>> {
+    match task {
+        Some(handle) => handle.await.unwrap(),
+        None => match AlwaysPending.await {},
+    }
+}
+
+/// A future that will never resolve
+struct AlwaysPending;
+
+impl std::future::Future for AlwaysPending {
+    type Output = std::convert::Infallible;
+
+    fn poll(
+        self: std::pin::Pin<&mut Self>,
+        _cx: &mut std::task::Context<'_>,
+    ) -> std::task::Poll<Self::Output> {
+        std::task::Poll::Pending
+    }
+}
+
+/// Start the given task/command
+fn start_task(task: Task, command: &mut Command) -> Result<Child> {
+    command
+        .kill_on_drop(true)
+        .env("RUST_BACKTRACE", if task.backtrace { "1" } else { "0" })
+        .spawn()
+        .context("failed to launch command")
+}
+
+/// Send all lines in the process' output
+async fn execute_task(
+    mut child: Child,
+    with_stdout: bool,
+    line_sender: LineSender,
+) -> Result<Option<ExitStatus>> {
+    let stderr = child
+        .stderr
+        .take()
+        .ok_or_else(|| anyhow!("child missing stderr"))?;
+
+    let stderr_sender = line_sender.clone();
+    let stderr = stream_consumer(stderr, CommandStream::StdErr, stderr_sender);
+
+    let stdout = if with_stdout {
+        let stdout = child
+            .stdout
+            .take()
+            .ok_or_else(|| anyhow!("child missing stdout"))?;
+        let stdout_sender = line_sender.clone();
+        Some(stream_consumer(stdout, CommandStream::StdOut, stdout_sender))
+    } else {
+        None
+    };
+
+    // either we wait on both stdout and stderr concurrently, or just stderr.
+    if let Some(stdout) = stdout {
+        tokio::try_join!(stdout, stderr)?;
+    } else {
+        stderr.await?;
+    }
+
+    let status = match child.wait().await {
+        Err(_) => None,
+        Ok(status) => Some(status),
+    };
+
+    Ok(status)
+}
+
+/// Send all lines in the given stream to the sender.
+async fn stream_consumer(
+    stream: impl AsyncRead + Unpin,
+    origin: CommandStream,
+    line_sender: LineSender,
+) -> Result<()> {
+    let mut lines = BufReader::new(stream).lines();
+
+    while let Some(line) = lines.next_line().await? {
+        let response = CommandExecInfo::Line(CommandOutputLine {
+            content: TLine::from_tty(&line),
+            origin,
+        });
+        if let Err(_) = line_sender.send(response) {
+            return Err(anyhow!("channel closed"));
+        }
```

Comment:
this is block call in async function, this look bad.

Reply:
Agreed, I wanted to keep using the `std` channels so that other code wasn't affected. But we could probably swap to [`tokio::sync::mpsc`](https://docs.rs/tokio/latest/tokio/sync/mpsc/struct.Receiver.html#method.blocking_recv) where relevant instead.

Reply:
MMM it seem send is non blocking so all good https://doc.rust-lang.org/std/sync/mpsc/struct.Sender.html#method.send